modules/javafx.controls/src/main/java/javafx/scene/control/ListCell.java

```diff
@@ -540,8 +541,13 @@ private void updateEditing() {
         final int editIndex = list == null ? -1 : list.getEditingIndex();
         final boolean editing = isEditing();
 
-        // Check that the list is specified, and my index is not -1
-        if (index != -1 && list != null) {
+        if (index == -1 || list == null) {
+            if(editing) {
+                updateEditingIndex = false;
+                cancelEdit();
+                updateEditingIndex = true;
+            }
+        } else {
```

Comment:
exactly :) And now we have duplicated code blocks in a single method which can  be improved by extracting them. It's a bit a matter of judgement/style of the codebase: in this particular case we probably should because 

- it's the same across all cell implementations
- a slight issue turned up in [review](https://github.com/openjdk/jfx/pull/473#discussion_r617762020) of the exact same code in TableCell (should guarantee that the flag is reliably reset to true, no matter what happens in cancelEdit) - fixing that increases the length of the duplicated block. 

Whatever the final outcome, it should be consistent across cells. We might wait with this until the TableCell is finalized, then do the same here - what do you think?

Unrelated to extracting or not: the sequence should have the same code comment everywhere it appears.

Reply:
I've now rewritten the code. It's now way simpler and avoids duplicate code. I guess the different implementation in the different cells all have some subtle differences.

Then i guess we wait for the other review to finish. The change to use try finalize seems very very reasonable and important to me.

Reply:
> 
> 
> I guess the different implementation in the different cells all have some subtle differences.
> 

actually, I doubt that they are different - my guess: we could formally prove that the logic is equivalent :) Or the other way: if they are, we should understand the differences. On first look your variant seems okay, as well.

> Then i guess we wait for the other review to finish. The change to use try finalize seems very very reasonable and important to me.

so we agree - will wait with a final review until the other is decided. You might consider removing the debug sysouts before that :)

Reply:
Hm okay, when we assume the result of match in TreeCell is the same as index != editIndex, then they are identical.
And when we assume that cancelEdit doesn't have relevant side effects, which can easily happen due to various listeners ...

I've removed the accidentally commited print statements!